### ISSUE TITLE
Ensemble Based Active Learning

### ISSUE THREAD

**Issue by NobleKennamer** [ElementAI/baal#51]
Would this project be interested in ensemble based techniques for active learning? As seen in [1] ensembles can be an effective technique for deep active learning achieving competitive/superior performance to competing techniques (ie MC-Dropout). More generally they have been widely used in non-deep based approaches to active learning as well (see [2] chapter 3 for a review).

All of the heuristics this library has already implemented could immediately be used for ensemble predictions. 

I see that SWAG is on the road map and this could be a more general feature of different ensemble style approaches (bootstrap, FGE, SWAG, diversity encouraging, etc...)

If this is something you all would like to move forward with. I would be happy to implement it. I've been working on active learning and already have my own implementations. Since finding this library I am in the midst of porting my code over to it, since I like it so much.

My only concern is that when resource constrained it is difficult to hold all of the ensemble members in memory and can be expensive to move them in and out. In my own experiments I've been generating the members of the ensemble one at a time and predicting on all data points from the pool set and validation set deleting the member and start training the new member. This might be a bit awkward given the current implementation of the ModelWrapper and ActiveLearningLoop.

 ``` python
Model.train()
active_loop.set()
```

Essentially for the above workflow to still work it would require the below line to be already executed and cached by the model and the active_loop would use the cache.

``` python
 probs = self.get_probabilities(pool, **self.kwargs)
```

Thanks for your time. 

[1] http://openaccess.thecvf.com/content_cvpr_2018/papers/Beluch_The_Power_of_CVPR_2018_paper.pdf

[2] https://www.morganclaypool.com/doi/abs/10.2200/S00429ED1V01Y201207AIM018

**Comment by Dref360** [comment]
Hi!
Yes we want to add support for Ensembles and SWAG would be a good start. If you want to help, that would be fantastic news! 

It is true that the switching models per batch is  suboptimal.

Could we train all the models while saving the weights to disk and when predicting we can have a loop over all models? I am definitely open to suggestion as we want this to be useful to a wide range of people. 

Thank you for reaching out!

**Comment by Dref360** [comment]
Do you think that if BaaL was providing utils that would make Ensemble easier to use would be valuable?

Something like:
* `predict_on_dataset_from_models(models : List[Callable], dataset, ...)`
  * This would take a list of model (or functions that would load a model and predict on the dataset for each of them.

As for training, we can let the user do it, even though we can make this easier. 

I tried to find some documentation on Pytorch Lightning or Pytorch Ignite on how to use "large" ensembles, but I did not find anything relevant.

**Comment by vfdev-5** [comment]
> Pytorch Ignite on how to use "large" ensembles, but I did not find anything relevant.

@Dref360 this is rather simple with PyTorch-Ignite:
```python
from ignite.engine import Engine
from ignite.utils import convert_tensor

models  = {
    "m1": m1,
    "m2": m2,
    ...
}

def predict_fn(e, batch):
     output = {}
     batch = convert_tensor(batch, device=device)
     with torch.no_grad():
         for name, model in models.items():
             model.eval()
             output[name] = model(batch)
     return output

predictor = Engine(predict_fn)

@predictor.on(Events.ITERATION_COMPLETED)
def save_batch(_):
    y_preds = predictor.state.output
    for name, y_pred in y_preds.items():
         save_prediction(y_pred, name, ...)

predictor.run(dataloader)
```

**Comment by Dref360** [comment]
Thanks @vfdev-5 it helps!

I think an issue that we have here is that we may get large ensemble that needs to be loaded/unloaded from the gpu.

Should we make multiple Engines, one per model, and run them sequentially?
Something like this: 

```python
# Lazy loading of the model
engines = (Engine(predict_fn_factory(model_arch[i])) for i in range(num_models))

states = [engine.run(dataloader) for engine in engines]
```

**Comment by vfdev-5** [comment]
@Dref360 I think moving models to CPU -> GPU -> CPU can be simply done with the above example with 1 engine as Engine is agnostic of model's device:
```python
def predict_fn(e, batch):
     output = {}
     batch = convert_tensor(batch, device="cuda")
     with torch.no_grad():
         for name, model in models.items():
             model.eval().to("cuda")
             output[name] = model(batch).cpu()  # .cpu() if we can convert simply the output
             model.to("cpu")
     return output
```

What do you think ?

**Comment by Dref360** [comment]
You are right, is it too early to optimize the loading/unloading?

Quick test that I ran and I think it is not too bad:
```
from torchvision.models import vgg16
import torch

def test(model):
    model.eval()
    model = model.cuda()
    model(torch.randn(1, 3, 224, 224).to('cuda'))
    model = model.cpu()

model = vgg16()
%timeit test(model)
# 456 ms ± 24.2 ms per loop

```

**Comment by vfdev-5** [comment]
> is it too early to optimize the loading/unloading?

Probably, yes. But thinking more about that, it is not obvious how can we can simply optimize this part ... (maybe, move several models with `non_blocking=True`). Another point is that doing `model.cpu()` does not directly means that GPU memory is freed. Some cuda empty cache ops will be necessary... CUDA context creation can be done elsewhere (but we wont gain a lot with that).

**Comment by NobleKennamer** [comment]
So I think support for predicting on ensembles would definitely be useful. My thinking when I originally posted the issue was that you would want a solution that works with the ModelWrapper as well, which seems difficult with the current implementation. 

But if you left training the ensemble to the user the rest of the library seems good to go. Any extra functionality for predicting on ensembles in an efficient way would be a major plus.

This wouldn't be a general solution, but most deep ensemble models use the exact same architecture so would just loading and unloading different state_dicts into same instantiation get around the issue that `model.cpu()` does not actually free the memory?

I'll be working today on an ensemble example using baal minus the ModelWrapper. I'll post when I'm done.

Thanks for all the discussion!

**Comment by Dref360** [comment]
Sorry for the delay, we are on a really tight schedule for NeurIPS.

I made a proposal for Ensemble : https://github.com/ElementAI/baal/pull/58

I would appreciate if you could provide some feedback :) 

The idea would be to add SWAG and CSG-MCMC using this infrastructure.

Thank you!
